vowpalwabbit/cats_tree.cc

```diff
@@ -0,0 +1,398 @@
+              base.learn(ec, v.parent_id);
+              _binary_tree.nodes[v.parent_id].learn_count++;
+              base.predict(ec, v.parent_id);
+              VW_DBG(ec) << "tree_c: learn() after binary predict:" << scalar_pred_to_string(ec)
+                << ", local_action = " << (local_action) << std::endl;
+              float trained_action = (ec.pred.scalar < 0) ? LEFT : RIGHT;
+              if (trained_action == local_action)
+              {
+                cost_parent =
+                  (std::min)(cost_v, cost_w) * std::abs(ec.pred.scalar) + (std::max)(cost_v, cost_w) * (1 - std::abs(ec.pred.scalar));
+                VW_DBG(ec) << "tree_c: learn() ec.pred.scalar == local_action" << std::endl;
+              }
+              else
+              {
+                cost_parent =
+                  (std::max)(cost_v, cost_w) * std::abs(ec.pred.scalar) + (std::min)(cost_v, cost_w) * (1 - std::abs(ec.pred.scalar));
+                VW_DBG(ec) << "tree_c: learn() ec.pred.scalar != local_action" << std::endl;
+              }
+            }
+
+          }
+          if (i == 0)
+            a_parent_cost = cost_parent;
+          else
+            b_parent_cost = cost_parent;
+        }
+        _a = { nodes[_a.node_id].parent_id, a_parent_cost };
+        _b = { nodes[_b.node_id].parent_id, b_parent_cost };
+      }
+
+      ec.l = saved_label;
+      ec.weight = saved_weight;
+      ec.pred = saved_pred;
+    }
+
+    void cats_tree::set_trace_message(std::ostream* vw_ostream) { _trace_stream = vw_ostream; }
+
+    cats_tree::~cats_tree()
+    {
+      if (_trace_stream != nullptr)
+        (*_trace_stream) << tree_stats_to_string() << std::endl;
+    }
+
+    std::string cats_tree::tree_stats_to_string()
+    {
+      return _binary_tree.tree_stats_to_string();
+    }
+
+    void predict(cats_tree& ot, single_learner& base, example& ec)
+    {
+      VW_DBG(ec) << "tree_c: before tree.predict() " << multiclass_pred_to_string(ec) << features_to_string(ec)
+        << std::endl;
+      ec.pred.multiclass = ot.predict(base, ec);
+      VW_DBG(ec) << "tree_c: after tree.predict() " << multiclass_pred_to_string(ec) << features_to_string(ec)
+        << std::endl;
+    }
+
+    void learn(cats_tree& tree, single_learner& base, example& ec)
+    {
+      VW_DBG(ec) << "tree_c: before tree.learn() " << cb_label_to_string(ec) << features_to_string(ec) << std::endl;
+      tree.learn(base, ec);
+      VW_DBG(ec) << "tree_c: after tree.learn() " << cb_label_to_string(ec) << features_to_string(ec) << std::endl;
+    }
+
+    base_learner* setup(options_i& options, vw& all)
+    {
+      option_group_definition new_options("CATS Tree Options");
+      uint32_t num_actions; // = K = 2^D
+      uint32_t bandwidth; // = 2^h#
+      uint32_t scorer_flag;
+      new_options.add(make_option("cats_tree", num_actions).keep().help("CATS Tree with <k> labels"))
+        .add(make_option("scorer_option", scorer_flag)
```

Comment:
is this flag needed?

Reply:
Yes.  Why do you ask?

Reply:
1) its expressed as an int 
2) if you remove it and change the `if` statement on line 376 the following vw cmds would be equivalent  `--cats_tree 128 --scorer_option 1` == `--cats_tree 128 --binary`. The latter being more expressive. This actually might work right now straight out of the box.
3) are the other link functions supported? or only glf1? 
4) what is the expected behavior when this command is parsed `--cats_tree 128 --scorer_option 0 --link logistic`?.

Reply:
The intent was to have some control over the learner that's in each node of the tree.  It also seems to be the case that more than just these two options were contemplated.  If that's the case the choice of int seems reasonable.  I will clarify with Maryam.  A reduction author may have a preference on what to use as it's base learner.  While someone else can override the base learner, this might not be desirable.

Reply:
We discussed offline during the vw meeting and @olgavrou is preparing a change addressing this point. It is equivalent and there's no change actual behavior or extensibility. We should review any concerns on that pr.

Reply:
ok sounds good.

Reply:
Discussed alternative here: https://github.com/mmajzoubi/vowpal_wabbit/pull/66